src/Compilers/CSharp/Portable/Symbols/Synthesized/InjectedAttributeSymbol.cs

```diff
@@ -0,0 +1,37 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols
+{
+    /// <summary>
+    /// Injected attribute symbols are injected early in the compilation, and so can be referenced in source.
+    /// But we track their usage and only emit them if they are used.
+    /// Their method bodies are always compiled, in case we do need to emit them.
+    /// </summary>
+    internal abstract class InjectedAttributeSymbol : SynthesizedEmbeddedAttributeSymbolBase
+    {
+        // All the diagnostics involved in constructing this symbol will only be produced
+        // if the symbol is referenced and so ends up emitted.
+        // We collect all those diagnostics here.
+        internal readonly DiagnosticBag Diagnostics;
+
+        public InjectedAttributeSymbol(
+            AttributeDescription description,
+            NamespaceSymbol containingNamespace,
+            CSharpCompilation compilation,
+            Func<CSharpCompilation, NamedTypeSymbol, DiagnosticBag, ImmutableArray<MethodSymbol>> getConstructors,
+            DiagnosticBag diagnostics)
+            : base(description, containingNamespace, compilation, getConstructors, diagnostics)
+        {
+            Diagnostics = diagnostics;
```

Comment:
>Diagnostics = diagnostics [](start = 12, length = 25)

Why do we pass a bag? Shouldn't each symbol allocate its own bag?

Reply:
I need to pass that bag into the `base(...)` call. So this constructor cannot be responsible for creating the bag. 
I considered making this constructor private and add a static factory method. What do you think?

---
In reply to: [215460876](https://github.com/dotnet/roslyn/pull/29180#discussion_r215460876) [](ancestors = 215460876)

Reply:
> I considered making this constructor private and add a static factory method. What do you think?

We can make this constructor private and have public constructor that delegates to this one, passing new bag. This way it will be clear that no one can pass an arbitrary bag in. 

---
In reply to: [215461425](https://github.com/dotnet/roslyn/pull/29180#discussion_r215461425) [](ancestors = 215461425,215460876)